src/compiler/scala/tools/nsc/transform/patmat/MatchAnalysis.scala

```diff
@@ -95,58 +95,84 @@ trait TreeAndTypeAnalysis extends Debugging {
-            Some(
-              // We must only include subtypes of the parents that conform to `tp`.
-              // See neg/virtpatmat_exhaust_compound.scala for an example.
-              parentSubtypes flatMap (_.getOrElse(Nil)) filter (_ <:< tp)
-            )
-          else None
+            // We must only include subtypes of the parents that conform to `tp`.
+            // See neg/virtpatmat_exhaust_compound.scala for an example.
+            parentSubtypes map (_.filter(_ <:< tp))
+          }
+          else Nil
         // make sure it's not a primitive, else (5: Byte) match { case 5 => ... } sees no Byte
         case sym if sym.isSealed =>
-          val subclasses = debug.patmatResult(s"enum $sym sealed, subclasses")(
-            // symbols which are both sealed and abstract need not be covered themselves, because
-            // all of their children must be and they cannot otherwise be created.
-            sym.sealedDescendants.toList
-              sortBy (_.sealedSortName)
-              filterNot (x => x.isSealed && x.isAbstractClass && !isPrimitiveValueClass(x))
-          )
 
           val tpApprox = typer.infer.approximateAbstracts(tp)
           val pre = tpApprox.prefix
 
-          Some(debug.patmatResult(s"enum sealed tp=$tp, tpApprox=$tpApprox as") {
-          // valid subtypes are turned into checkable types, as we are entering the realm of the dynamic
-            subclasses flatMap { sym =>
+          def filterChildren(children: List[Symbol]): List[Type] = {
+            children flatMap { sym =>
               // have to filter out children which cannot match: see ticket #3683 for an example
               // compare to the fully known type `tp` (modulo abstract types),
               // so that we can rule out stuff like: sealed trait X[T]; class XInt extends X[Int] --> XInt not valid when enumerating X[String]
               // however, must approximate abstract types in
 
-              val memberType  = nestedMemberType(sym, pre, tpApprox.typeSymbol.owner)
-              val subTp       = appliedType(memberType, sym.typeParams.map(_ => WildcardType))
+              val memberType = nestedMemberType(sym, pre, tpApprox.typeSymbol.owner)
+              val subTp = appliedType(memberType, sym.typeParams.map(_ => WildcardType))
               val subTpApprox = typer.infer.approximateAbstracts(subTp) // TODO: needed?
               // debug.patmat("subtp"+(subTpApprox <:< tpApprox, subTpApprox, tpApprox))
               if (subTpApprox <:< tpApprox) Some(checkableType(subTp))
               else None
             }
-          })
+          }
+
+          if(grouped) {
+            def enumerateChildren(sym: Symbol) = {
+              sym.children.toList
+                .sortBy(_.sealedSortName)
+                .filterNot(x => x.isSealed && x.isAbstractClass && !isPrimitiveValueClass(x))
+            }
+
+            // enumerate only direct subclasses,
```

Comment:
It can be fooled by such an example actually:
```Scala
object Test {

  sealed class C

  trait A extends C

  trait B extends C

  trait AB extends A with B

  def `A missing`(c: C) = c match {
    case _: B => ???
  }

  def exhaustive(c: C) = c match {
    case _: AB => ???
  }
}
```

The second match should report only `C()` but it claims `A` and `B` are missing too:

```
~/Documents/scala-2.11/sandbox:rebase-4379$ ../qbin/scalac  -Xfatal-warnings -unchecked -Xfatal-warnings Test.scala 
Test.scala:11: warning: match may not be exhaustive.
It would fail on the following inputs: (_ : Test.A), C()
  def `A missing`(c: C) = c match {
                          ^
Test.scala:15: warning: match may not be exhaustive.
It would fail on the following inputs: (_ : Test.A), (_ : Test.B), C()
  def exhaustive(c: C) = c match {
                         ^
error: No warnings can be incurred under -Xfatal-warnings.
two warnings found
one error found
```

However, I checked against Scala 2.11.2 and the output is exactly the same... I can try to fix it...